numpy/core/src/umath/override.c

```diff
@@ -0,0 +1,479 @@
+    else if (strcmp(method, "reduceat") == 0) {
+        status = normalize_reduceat_args(ufunc, args, &normal_args,
+                                         &normal_kwds);
+    }
+    /* ufunc.outer */
+    else if (strcmp(method, "outer") == 0) {
+        status = normalize_outer_args(ufunc, args, &normal_args, &normal_kwds);
+    }
+    /* ufunc.at */
+    else if (strcmp(method, "at") == 0) {
+        status = normalize_at_args(ufunc, args, &normal_args, &normal_kwds);
+    }
+    /* unknown method */
+    else {
+        PyErr_Format(PyExc_TypeError,
+                     "Internal Numpy error: unknown ufunc method '%s' in call "
+                     "to PyUFunc_CheckOverride", method);
+        status = -1;
+    }
+    if (status != 0) {
+        Py_XDECREF(normal_args);
+        goto fail;
+    }
+
+    len = PyTuple_GET_SIZE(normal_args);
+    override_args = PyTuple_New(len + 2);
+    if (override_args == NULL) {
+        goto fail;
+    }
+
+    Py_INCREF(ufunc);
+    /* PyTuple_SET_ITEM steals reference */
+    PyTuple_SET_ITEM(override_args, 0, (PyObject *)ufunc);
+    method_name = PyUString_FromString(method);
+    if (method_name == NULL) {
+        goto fail;
+    }
+    Py_INCREF(method_name);
+    PyTuple_SET_ITEM(override_args, 1, method_name);
+    for (i = 0; i < len; i++) {
+        PyObject *item = PyTuple_GET_ITEM(normal_args, i);
+
+        Py_INCREF(item);
+        PyTuple_SET_ITEM(override_args, i + 2, item);
+    }
+    Py_DECREF(normal_args);
+
+    /* Call __array_ufunc__ functions in correct order */
+    while (1) {
+        PyObject *array_ufunc;
+        PyObject *override_obj;
+
+        override_obj = NULL;
+        *result = NULL;
+
+        /* Choose an overriding argument */
+        for (i = 0; i < noa; i++) {
+            obj = with_override[i];
+            if (obj == NULL) {
+                continue;
+            }
+
+            /* Get the first instance of an overriding arg.*/
+            override_obj = obj;
+
+            /* Check for sub-types to the right of obj. */
+            for (j = i + 1; j < noa; j++) {
+                other_obj = with_override[j];
+                if (other_obj != NULL &&
+                    PyObject_Type(other_obj) != PyObject_Type(obj) &&
+                    PyObject_IsInstance(other_obj,
+                                        PyObject_Type(override_obj))) {
+                    override_obj = NULL;
+                    break;
+                }
+            }
+
+            /* override_obj had no subtypes to the right. */
+            if (override_obj) {
+                /* We won't call this one again */
+                with_override[i] = NULL;
+                break;
+            }
+        }
+
+        /* Check if there is a method left to call */
+        if (!override_obj) {
+            /* No acceptable override found. */
+            PyErr_SetString(PyExc_TypeError,
+                            "__array_ufunc__ not implemented for this type.");
```

Comment:
My experimentation on https://github.com/charris/numpy/pull/19 just turned up this error message like this, which I did not find very user friendly. Can we change it to include the ufunc name and input types and/or reprs, maybe

    __array_ufunc__ not implemented for method '__call__' on ufunc <ufunc 'add'>
    with input types <MyArray> and list.

It would be nice to include the output, too, if present.